### PR TITLE
DR-2052 Add retry on exclusive dataset lock on snapshot create

### PR DESCRIPTION
**Pull request by nmalfroy**
10 minutes exponential backoff (e.g. default backoff)

### CODE REVIEW

File: src/main/java/bio/terra/service/dataset/flight/LockDatasetStep.java [hunk 1]
```diff
@@ -62,8 +62,9 @@ public StepResult doStep(FlightContext context) {
       } else {
         return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, notFoundEx);
       }
-    } catch (DatasetLockException ex) {
-      return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, ex);
+    } catch (DatasetLockException e) {
+      logger.warn("Dataset could not be locked.  Retrying", e);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/dataset/flight/LockDatasetStep.java]: It is necessary to log here? Stairway already logs every retry event with `info`

https://github.com/DataBiosphere/stairway/blob/develop/stairway/src/main/java/bio/terra/stairway/Flight.java#L329

File: src/main/java/bio/terra/service/dataset/flight/LockDatasetStep.java [hunk 2]
```diff
@@ -62,8 +62,9 @@ public StepResult doStep(FlightContext context) {
...
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
     } catch (RetryQueryException retryQueryException) {
       return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/dataset/flight/LockDatasetStep.java]: Any idea why this `StepResult()` doesn't pass the caught exception? I'm wondering why it wouldn't be done here but is done in the new code.

**Review by pshapiro4broad — APPROVED** [review]
Looks OK, may be more logging than needed.

File: src/main/java/bio/terra/service/dataset/flight/LockDatasetStep.java (hunk 1 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/flight/LockDatasetStep.java]: Fair enough...I'll remove that

File: src/main/java/bio/terra/service/dataset/flight/LockDatasetStep.java (hunk 2 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/flight/LockDatasetStep.java]: It's a good question we don't seem to be consistent about when we include the exception.  Lemme see if I can figure out what get's logged

**Review by samanehsan — APPROVED** [review]
Lgtm :octocat:

**Pull request merged**